### PR TITLE
Iteratorsize for partition iterator

### PR DESCRIPTION
**Pull request by mschauer**
Before, collect of a partition could fail because it assumed the existence of a length method via the iteratorsize fall back.

### CODE REVIEW

File: test/functional.jl [hunk 1]
```diff
@@ -264,10 +264,10 @@ let v = collect(Base.partition([1,2,3,4,5], 1))
     @test all(i->v[i][1] == i, v)
 end
 
-let v = collect(Base.partition([1,2,3,4,5], 2))
-    @test v[1] == [1,2]
-    @test v[2] == [3,4]
-    @test v[3] == [5]
+let v1 = collect(Base.partition([1,2,3,4,5], 2)), v2 = collect(Base.partition(Base.flatten([1,2,3,4,5]), 2))
+    @test v1[1] == v1[1] == [1,2]
```
Comment (tkelman) [inline test/functional.jl]: presumably this should be `v1[1] == v2[1]` ?

**Comment by mschauer** [comment]
The test failure after adressing @tkelman's spot is unrelated, previously passing test here https://ci.appveyor.com/project/JuliaLang/julia/build/1.0.2015

**Comment by mschauer** [comment]
Personal assessment: this change is straight forward.

**Comment by tkelman** [comment]
Thanks for the bump, I agree this looks fine.

I think all the iterator size business still needs more thorough interface documentation, right?

**Comment by mschauer** [comment]
Yes, I wrote https://gist.github.com/mschauer/c8d8b7eb5b455cb12ddc9bda15695203 but I would like to wait for feedback before writing more docs. See issue #15977

**Comment by mschauer** [comment]
Bump.

File: base/iterator.jl [hunk 2]
```diff
@@ -490,6 +490,9 @@ type PartitionIterator{T}
 end
 
 eltype{T}(::Type{PartitionIterator{T}}) = Vector{eltype(T)}
+partition_iteratorsize(::HasShape) = HasLength()
+partition_iteratorsize{T}(isz::T) = isz
```
Comment (rfourquet) [inline base/iterator.jl]: The type parameter can be omitted.

File: base/iterator.jl [hunk 3]
```diff
@@ -490,6 +490,9 @@ type PartitionIterator{T}
...
+iteratorsize{T}(P::Type{PartitionIterator{T}}) = partition_iteratorsize(iteratorsize(T))
```
Comment (rfourquet) [inline base/iterator.jl]: the argument `P` can be omitted, and when rebasing update to the `where` syntax

File: test/functional.jl [hunk 4]
```diff
@@ -264,10 +264,10 @@ let v = collect(Base.partition([1,2,3,4,5], 1))
     @test all(i->v[i][1] == i, v)
 end
 
-let v = collect(Base.partition([1,2,3,4,5], 2))
-    @test v[1] == [1,2]
-    @test v[2] == [3,4]
-    @test v[3] == [5]
+let v1 = collect(Base.partition([1,2,3,4,5], 2)), v2 = collect(Base.partition(Base.flatten([1,2,3,4,5]), 2))
```
Comment (rfourquet) [inline test/functional.jl]: wrap at 92 (IIRC) chars

**Review by rfourquet — CHANGES_REQUESTED** [review]
LGTM! (modulo the few comments)

**Review by Sacha0 — APPROVED** [review]
lgtm! Thanks @mschauer! :)

(The CI failures appear unrelated.)

**Comment by StefanKarpinski** [comment]
If this is good to go, please merge, @rfourquet since you have some outstanding review comments.

**Comment by mschauer** [comment]
@rfourquet Let me make one amendment first, after #22691 ``partition(flatten([1,2,3,4,5]),2)`` will actually have a length and not trigger anymore the ``SizeUnknown``-code path of collect. One can replace with``partition(flatten([[1,2],[3,4],5]),2)``.

File: base/iterators.jl [hunk 5]
```diff
@@ -799,6 +799,11 @@ mutable struct PartitionIterator{T}
 end
 
 eltype(::Type{PartitionIterator{T}}) where {T} = Vector{eltype(T)}
+partition_iteratorsize(::HasShape) = HasLength()
+partition_iteratorsize(isz) = isz
+function iteratorsize(::Type{PartitionIterator{T}}) where {T}
+    partition_iteratorsize(iteratorsize(T))
+end
```
Comment (rfourquet) [inline base/iterators.jl]: out of curiosity: you switched to the long-form function syntax because of the `where` syntax?

**Comment by rfourquet** [comment]
Thanks @mschauer !

File: base/iterators.jl (hunk 5 above)
Comment (mschauer) [inline base/iterators.jl]: yes

**Pull request merged**